j2cl-tasks/src/main/java/com/vertispan/j2cl/build/provided/TurbineTask.java

```diff
@@ -0,0 +1,173 @@
+
+    @Override
+    public String getOutputType() {
+        return OutputTypes.STRIPPED_BYTECODE_HEADERS;
+    }
+
+    @Override
+    public String getTaskName() {
+        return "turbine";
+    }
+
+    @Override
+    public String getVersion() {
+        return "0";
+    }
+
+    @Override
+    public Task resolve(Project project, Config config) {
+        int version = getJavaVersion();
+        if(version == 8) {
+            return super.resolve(project, config);
+        }
+
+        // emits only stripped bytecode, so we're not worried about anything other than .java files to compile and .class on the classpath
+        Input ownSources = input(project, OutputTypes.STRIPPED_SOURCES).filter(JAVA_SOURCES);
+
+        List<File> extraClasspath = config.getExtraClasspath();
+
+        List<Input> compileClasspath = scope(project.getDependencies(), Dependency.Scope.COMPILE).stream()
+                .map(p -> input(p, OutputTypes.STRIPPED_BYTECODE_HEADERS))
+                .map(input -> input.filter(JAVA_BYTECODE))
+                .collect(Collectors.toList());
+
+        return context -> {
+
+            Set<String> deps = Stream.concat(extraClasspath.stream().map(File::toString),
+                            compileClasspath.stream().map(Input::getParentPaths).flatMap(Collection::stream).map(Path::toFile)
+                                    .map(File::toString)
+                                    .map(path -> path + "/output.jar"))
+                    .collect(Collectors.toSet());
+
+            File resultFolder = context.outputPath().toFile();
+            File output = new File(resultFolder, "output.jar");
+
+            List<String> sources = ownSources.getFilesAndHashes()
+                    .stream()
+                    .map(p -> SourceUtils.FileInfo.create(p.getAbsolutePath().toString(), p.getSourcePath().toString()))
+                    .map(SourceUtils.FileInfo::sourcePath)
+                    .collect(Collectors.toList());
+
+            try {
+                Main.Result result = Main.compile(
+                        optionsWithBootclasspath()
+                                .setSources(ImmutableList.copyOf(sources))
+                                .setOutput(output.toString())
+                                .setClassPath(ImmutableList.copyOf(deps))
+                                //TODO take a look at ReducedClasspathMode
+                                //.setReducedClasspathMode(TurbineOptions.ReducedClasspathMode.JAVABUILDER_REDUCED)
+                                .build());
+
+                System.out.println("turbine finished: " + result);
+                extractJar(output.toString(), resultFolder.toString());
+            } catch (TurbineError e) {
+                //ohhhh apt is in maven reactor
+                System.out.println(e.getMessage());
+            }
+        };
+    }
+
+    public void extractJar(String zipFilePath, String extractDirectory) {
```

Comment:
Take a quick look at 22319e6d4be9c01c4edd2ed82748731607b97e96? I used that zip unpacking code originally to deal with j2cl tools that didnt know how to use the filesystem directly. That would let us drop the new commons-compress dependency.